src/main/java/org/spongepowered/common/event/tracking/phase/generation/GeneralGenerationPhaseState.java

```diff
@@ -134,16 +134,20 @@ public final void unwind(G context) {
     public boolean spawnEntityOrCapture(G context, Entity entity, int chunkX, int chunkZ) {
         final ArrayList<Entity> entities = new ArrayList<>(1);
         entities.add(entity);
-        final SpawnEntityEvent event = SpongeEventFactory.createSpawnEntityEventSpawner(Sponge.getCauseStackManager().getCurrentCause(),
-            entities);
-        SpongeImpl.postEvent(event);
-        if (!event.isCancelled() && event.getEntities().size() > 0) {
-            for (Entity item : event.getEntities()) {
-                ((IMixinWorldServer) item.getWorld()).forceSpawnEntity(item);
+        try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+            frame.addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.WORLD_SPAWNER);
```

Comment:
Should be adding the world to the cause here, in the frame.

Reply:
Or, you could be pushing the world to the cause stack the moment the phase is switched (in the context of when world generation phase states are switched on).

Reply:
@gabizou 

Thanks for the feedback. Based on your comments, I'd need to push the world into the stack...here?

https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/mixin/core/world/MixinWorldEntitySpawner.java#L344

Reply:
You could, but if you’re pushing it outside of a frame, like that injection point, you need to pop that cause off at the return point. If you do it here where you are throwing an event, you can ignore having to pop the cause off the stack since the frame closing will do that for you.